[autograd] fix allow_unused checking for C++ API

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34035 [autograd] fix allow_unused checking for C++ API**

Bug for the conditon check in https://github.com/pytorch/pytorch/pull/24342, realized we don't have tests in either
python or cpp to catch this, so added testes for both python and cpp.

Thanks @hczhu on capturing it!

Differential Revision: [D20198837](https://our.internmc.facebook.com/intern/diff/D20198837)

wanchaol committed:
[autograd] fix allow_unused checking for C++ API

Bug for the conditon check, realized we don't have tests in either
python or cpp to catch this, so added testes for both python and cpp.

Thanks @hczhu on capturing it!

[ghstack-poisoned]

hczhu commented on torch/csrc/autograd/autograd.cpp:
```diff
@@ -100,7 +100,7 @@ variable_list run_backward(
       roots, grad_outputs, keep_graph, create_graph, output_edges);
   // check if grad_inputs contains None or not base on the allow_unused flag
-  if (inputs.empty()) {
+  if (!inputs.empty()) {
```
nit:
skip the check if `allow_unused` is true.
```
if (!inputs.empty() && !allow_unused) {

```

albanD commented:
@wanchaol this is BC-breaking change right? This will raise errors for user code that used to run.

wanchaol commented:
> @wanchaol this is BC-breaking change right? This will raise errors for user code that used to run.

@albanD hmm yeah, but this is only BC-breaking for the autograd C++ API (not the one from python as they are not sharing the logic, and C++ API did it wrong for this check). Also I would say if user are using it that way it should be wrong usage at first (i.e. the same thing will throw error when using python).

albanD reviewed (approved):
Same nit as the one above.
I got confused by the fact that you added a test for the python code. But it was already working. Sounds good.

wanchaol committed:
Update on "[autograd] fix allow_unused checking for C++ API"


Bug for the conditon check in https://github.com/pytorch/pytorch/pull/24342, realized we don't have tests in either
python or cpp to catch this, so added testes for both python and cpp.

Thanks @hczhu on capturing it!

[ghstack-poisoned]
